SSA/Experimental/Bits/Fast/BitStream.lean

```diff
@@ -367,28 +367,60 @@ variable {w : Nat} {x y : BitVec w} {a b a' b' : BitStream}
 
 local infix:20 " ≈ʷ " => EqualUpTo w
 
-theorem xor_xor_eq_not {a b : Bool} : xor (!xor a b) b = !a := by
-  cases a
-  <;> cases b
-  <;> simp
-
-theorem xor_and_eq_and {a b : Bool} : (!xor a b && b) = (a && b) := by
-  cases a
-  <;> cases b
-  <;> simp
-
 theorem neg_neg : a = - - a := by
   ext i
   have neg_lemma :
     a.neg.negAux i = ⟨a i, (a.negAux i).2⟩ := by
     induction' i with i ih
     · simp [neg, negAux]
-    · simp [neg, negAux, ih, xor_xor_eq_not, xor_and_eq_and]
+    · simp [neg, negAux, ih, Bool.xor_xor_eq_not, Bool.xor_and_eq_and]
   simp [Neg.neg, neg, neg_lemma]
 
--- TODO: This sorry is difficult, and will be proven in a later Pull Request.
-theorem ofBitVec_sub : ofBitVec (x - y) ≈ʷ (ofBitVec x) - (ofBitVec y)  := by
-  sorry
+def subCarries? (a b : BitStream) (i : Nat) : Bool :=
+  let carry : Bool := match i with
+  | 0 => false
+  | i + 1 => a.subCarries? b i
+  (!a i && b i || !xor (a i) (b i) && carry)
+
+theorem neg_or_add (i : Nat) :
+    (b.negAux i).2 = false ∨ (a.addAux (fun n => (b.negAux n).1) i).2 = false := by
+  induction' i with i ih
+  · simp [negAux, addAux, BitVec.adcb]
+    cases b 0
+    <;> simp
+  · simp [negAux, addAux, BitVec.adcb]
+    cases' ih with l r
+    · simp [l]
+    · cases b (i + 1)
+      <;> cases a (i + 1)
+      <;> simp [r]
+
+theorem subCarries?_correct (i : Nat) :
+    a.subCarries? b i = xor (b (i + 1)) ((!b (i + 1)) != ((b.negAux i).2 != (a.addAux (fun n => (b.negAux n).1) i).2)) := by
+  induction' i with i ih
+  · by_cases af : a 0
+    <;> simp [subCarries?, negAux, addAux, BitVec.adcb, af]
+  · simp [subCarries?, ih, negAux, addAux, BitVec.adcb]
+    cases a (i + 1)
+    <;> cases b (i + 1)
+    <;> cases b (i + 2)
+    <;> cases' @neg_or_add a b i with h h
+    <;> simp [h]
+
+theorem carry_succ_left {a b : BitStream} (i : ℕ) :
+    xor (b (i + 1)) (a.subCarries? b i) = ((!b (i + 1)) != ((b.negAux i).2 != (a.addAux (fun n => (b.negAux n).1) i).2)) := by
+  simp [Bool.xor_shift, subCarries?_correct i]
+
+theorem sub_add_lemma (i : Nat) :
+    a.subAux b i = ⟨(a.addAux b.neg i).1, subCarries? a b i⟩  := by
+  induction' i with i ih
+  · simp [subAux,addAux,negAux, BitVec.adcb, subCarries?, neg]
+  · simp [subAux,addAux,negAux, BitVec.adcb, ih, carry_succ_left i, subCarries?, neg]
+    rfl
+
+theorem sub_add_neg {a b : BitStream} : a - b = a + (-b) := by
```

Comment:
```suggestion
theorem sub_eq_add_neg {a b : BitStream} : a - b = a + (-b) := by
```

Reply:
Maybe check in lean if there is a naming example, e.g., for Nat or sth similar.

Reply:
Yes, there is 
```lean
theorem Int.sub_eq_add_neg {a : Int} {b : Int} : a - b = a + -b	
```